scripts/run_forge.sh

```diff
@@ -0,0 +1,160 @@
+#!/bin/bash
+
+# Copyright (c) Aptos
+# SPDX-License-Identifier: Apache-2.0
+
+#
+# This script runs Forge using the current configured kubectl context
+# It is designed to be invoked within Github Actions, but can be run locally
+# if the necessary environment variables are set.
+#
+
+# for calculating regression
+TPS_THRESHOLD=1500
+P99_LATENCY_MS_THRESHOLD=5000
+
+FORGE_OUTPUT=${FORGE_OUTPUT:-forge_output.txt}
+FORGE_REPORT=${FORGE_REPORT:-forge_report.json}
+AWS_ACCOUNT_NUM=${AWS_ACCOUNT_NUM:-$(aws sts get-caller-identity | jq -r .Account)}
+AWS_REGION=${AWS_REGION:-us-west-2}
+
+# forge test runner customization
+FORGE_RUNNER_MODE=${FORGE_RUNNER_MODE:-k8s}
+FORGE_NAMESPACE_KEEP=${FORGE_NAMESPACE_KEEP:-false}
+FORGE_ENABLE_HAPROXY=${FORGE_ENABLE_HAPROXY:-true}
+
+if [ -z "$FORGE_NAMESPACE" ]; then
+    new_namespace="forge-${USER}-$(date '+%s')"
+    echo "FORGE_NAMESPACE not set, using auto-generated namespace: ${new_namespace}"
+fi
+FORGE_NAMESPACE=${FORGE_NAMESPACE:-$new_namespace}
+# clean up namespace name
+# replace non alphanumeric chars with dash
+FORGE_NAMESPACE="${FORGE_NAMESPACE//[^[:alnum:]]/-}"
+# use the first 64 chars only for namespace, as it is the maximum for k8s resources
+FORGE_NAMESPACE=${FORGE_NAMESPACE:0:64}
+
+
+if [ "$FORGE_NAMESPACE_KEEP" = "true" ]; then
+    KEEP_ARGS="--keep"
+fi
+
+if [ "$FORGE_ENABLE_HAPROXY" = "true" ]; then
+    ENABLE_HAPROXY_ARGS="--enable-haproxy"
+fi
+
+
+if [ -z "$IMAGE_TAG" ]; then
+    echo "IMAGE_TAG not set"
+    exit 1
+fi
+
+echo "Ensure image exists"
+img=$(aws ecr describe-images --repository-name="aptos/validator" --image-ids=imageTag=$IMAGE_TAG)
+if [ $? != 0 ]; then
+    echo "IMAGE_TAG does not exist: ${IMAGE_TAG}"
+    exit 1
+fi
+
+FORGE_CLUSTER_NAME=$(kubectl config current-context | grep -oE 'aptos.*')
+
+echo "Using cluster ${FORGE_CLUSTER_NAME} from current kubectl context"
+
+FORGE_START_TIME_MS="$(date '+%s')000"
+
+# # Run forge with test runner in k8s
+if [ "$FORGE_RUNNER_MODE" = "local" ]; then
+
+    cargo run -p forge-cli -- test k8s-swarm \
+        --image-tag $IMAGE_TAG \
+        --namespace $FORGE_NAMESPACE \
+        --port-forward $KEEP_ARGS $ENABLE_HAPROXY_ARGS | tee $FORGE_OUTPUT
+
+    FORGE_EXIT_CODE=$?
+
+    # try to kill orphaned port-forwards
+    if [ -z "$KEEP_ARGS" ]; then
+        ps -A | grep  "kubectl port-forward -n $FORGE_NAMESPACE" | awk '{ print $1 }' | xargs -I{} kill -9 {}
+    fi
+
+else
+    # try deleting existing forge pod of same name
+    # since forge test runner will run in a pod that matches the namespace
+    # this will pre-empt the existing forge test in the same namespace and ensures
+    # we do not have any dangling test runners
+    FORGE_POD_NAME=$FORGE_NAMESPACE
+    kubectl delete pod $FORGE_POD_NAME || true
```

Comment:
What will happen if there is another test running under same user (because of multiple PRs) ?

Reply:
If you look at the way FORGE_NAMESPACE is set in the Github Action, you'll see that it creases 1 forge namespace/pod per PR. If you trigger Forge many times on the same PR, each time it will preempt the last run.

Reply:
If FORGE_NAMESPACE is not provided, a unique namespace will be used such that there will be no collision